src/formatters/default_formatter.erl

```diff
@@ -1355,24 +1349,55 @@ lay_application(Name, Arguments, SpacesWithinParentheses, Ctxt) ->
     end.
 
-%% @doc Might produce a new AST on which the functions in the export list
-%%      are sorted alphabetically if 'sort_function_exports' was set to 'true'
-sort_function_exports([Arguments0], SortFun) ->
-    case erl_syntax:subtrees(Arguments0) of
-        [] ->
-            %% node was a leaf node, skip
-            [Arguments0];
-        [SubTrees0] ->
-            SubTrees1 = lists:sort(SortFun, SubTrees0),
-            Arguments1 = erl_syntax:update_tree(Arguments0, [SubTrees1]),
-            [Arguments1]
-    end;
-sort_function_exports(Arguments, _SortFun) ->
-    Arguments.
+%% @doc Might produce a new AST node if the following criteria is met:
+%%        1. we know that the AST node is ONLY a list of arity qualifiers
+%%           (e.g.: [name/0]), and nothing else (no other kind of AST nodes
+%%           in the list)
+%%        2. the 'sort_arity_qualifiers' option was set to true
+%%      In short, we only will sort arity qualifiers present in the '-export',
+%%      '-export_type', and '-optional_callbacks' attributes.
+maybe_sort_arity_qualifiers(OriginalAST, Ctxt) ->
+    %% The 'OriginalAST' variable is a #list{} AST node for export lists,
+    %% and a list of #tree{} AST nodes for both export_type and optional_callbacks
+    %% lists. In order to unify the code handling the sorting for any of the cases,
+    %% we have to wrap the list of #tree{} AST nodes in a #list{} AST node, sort it,
+    %% and then unwrap it.
+    case Ctxt#ctxt.sort_arity_qualifiers_match andalso Ctxt#ctxt.sort_arity_qualifiers of
+        false ->
+            OriginalAST;
+        true when is_list(OriginalAST) ->
+            [UnwrappedAST] = OriginalAST,
+            do_sort_arity_qualifiers(UnwrappedAST);
+        true ->
+            ArityQualifiers = erl_syntax:list_elements(OriginalAST),
+            SortedArityQualifiers = do_sort_arity_qualifiers(ArityQualifiers),
+            erl_syntax:update_tree(OriginalAST, [SortedArityQualifiers])
+    end.
+
+%% @doc Might produce a new AST on which the arity qualifiers are sorted
+%%      alphabetically if 'sort_arity_qualifiers' was set to 'true'.
+%%      These arity qualifiers are the items in the export lists, export_type lists,
+%%      and other module attributes that contain function references in the form of
+%%      '[fun1/0, fun2/1]'.
+do_sort_arity_qualifiers(Arguments0) ->
+    case is_list(Arguments0) of
+        true ->
+            [Sorted] = do_sort_arity_qualifiers(erl_syntax:list(Arguments0)),
+            erl_syntax:list_elements(Sorted);
+        false ->
+            case erl_syntax:subtrees(Arguments0) of
+                [] ->
+                    %% node was a leaf node, skip
+                    [Arguments0];
+                [SubTrees0] ->
+                    SubTrees1 = lists:sort(fun sort_arity_qualifiers_alphabetically/2, SubTrees0),
+                    [erl_syntax:update_tree(Arguments0, [SubTrees1])]
+            end
+    end.
```

Comment:
```suggestion
do_sort_arity_qualifiers(Arguments0) when is_list(Arguments0) ->
        [Sorted] = do_sort_arity_qualifiers(erl_syntax:list(Arguments0)),
        erl_syntax:list_elements(Sorted);
do_sort_arity_qualifiers(Arguments0) ->
        case erl_syntax:subtrees(Arguments0) of
            [] ->
                %% node was a leaf node, skip
                [Arguments0];
            [SubTrees0] ->
                SubTrees1 = lists:sort(fun sort_arity_qualifiers_alphabetically/2, SubTrees0),
                [erl_syntax:update_tree(Arguments0, [SubTrees1])]
        end.
```

Reply:
Hehe, I very much approve of that refactor, which is how things used to be at first, before the comment about putting everything together in a single function clause; way more readable this way :D